### PR TITLE
More parallelism

### PR DESCRIPTION
**Pull request by danking**

### CODE REVIEW

**Comment by danking** [comment]
Deploy commits don't need to cleanup which adds some latency to this PR. We should probably use xargs -P4 to delete instances 4-way parallel. This PR is ~46 minutes, including all the cleanup time, where as deploys are 46 minutes *without the cleanup time*. Notice two things: (1) the service backend is again the critical path (2) some local backend tests took quite a while to get scheduled. Seems fishy to me that it took ~16 minutes to find a core for the local backend tests to run on.

Anyway, seems good to use more fine-grained parallelism. This should help keep the cluster large-ish and turning over fast so that users get a great experience during the work day.

---

A deploy commit:

Screen Shot 2023-05-17 at 17 30 55

This PR:

Screen Shot 2023-05-17 at 17 34 40

**Comment by danking** [comment]
Probably we should iterate on this PR, accumulating speed boosts until we get a satisfying PR latency.

**Comment by danking** [comment]
down to 38 minutes, with the cleaning up GCP instances now a significant delay.

Screen Shot 2023-05-17 at 23 02 20

**Comment by danking** [comment]
This was 11257f9473bfc17fd4db0ee6b6164da300b382af:

Screen Shot 2023-05-18 at 11 50 02

**Comment by danking** [comment]
`bbfc2a1e72822328bf5c8fd33d0e6f4ce5692da3`:
Screen Shot 2023-05-18 at 13 28 28

**Comment by danking** [comment]
Latest changes are worse e1699ab16df88fdc8d44099defe319d9f60f1ab0

Screen Shot 2023-05-18 at 20 25 44

**Comment by danking** [comment]
This looks to me like we just have bad long tails. If we can 90% and 99% test times down further maybe we're basically good. Seems feasible for the entire service backend tests to take <20 minutes.

**Comment by danking** [comment]
bd4e923ffdf13e847dab5017f78a3a1716d31c51 (36 minutes):

Screen Shot 2023-05-19 at 10 49 01

**Comment by danking** [comment]
fc0d048c0c07c9ec61ff186bc0a117333cd3054e (40 minutes):

Screen Shot 2023-05-19 at 17 27 32

**Comment by danking** [comment]
2079a85bce (41 minutes):

I could try to make the tests even more fine-grained and split up even more long-running tests. Seems like some of the bottlenecks I'm hitting now are:
1. Introducing an image with the wheel already installed isn't worthwhile, it adds 2.5 min latency.
2. The large number of splits often requires default Hail to scale up adding a 2min delay (It would be great to get that down).


I'm gonna revert the change that added images and maybe try to reduce service backend parallelism a bit. 36 minutes is an improvement. We should probably focus on making Hail faster rather than trying to squeeze lower latency out of parallelism.

Screen Shot 2023-05-22 at 12 30 47
Screen Shot 2023-05-22 at 12 31 23

**Comment by danking** [comment]
Rate limiting is a bit too restrictive methinks. That said, there are a lot of deadlocks when the rate limits kick in. Perhaps worth diving more deeply into the deadlocks at some point. The critical path service backend test is #35. Here are its slowest ones. I think we should probably checkpoint the import VCF, but, regardless, pc_relate just needs to be made faster.

I know how to double the speed of PCA. It's in a dead branch of mine. Patrick will incorporate those ideas into his rewrite using the new math.

If #35 was the same speed as the next slowest, we'd save 3 minutes. I think we can save ~7 minutes by cutting all these slow tests down so that the distribution of runtimes is more uniform.

```
256.68s call     hail/methods/relatedness/test_pc_relate.py::test_pc_relate_against_R_truth
178.28s call     hail/methods/test_pca.py::test_spectra_2[triplet0]
102.60s call     hail/vds/test_vds.py::test_truncate_reference_blocks
82.86s call     hail/backend/test_service_backend.py::test_tiny_driver_has_tiny_memory
```

f1ac37dbeb3625cbf91f1f9df5399f3723843029 (40 minutes) (https://ci.hail.is/batches/7484187):

Screen Shot 2023-05-25 at 12 01 55
Screen Shot 2023-05-25 at 12 01 28
Screen Shot 2023-05-25 at 12 01 07
Screen Shot 2023-05-25 at 12 01 03

**Comment by danking** [comment]
Also! Chris' zstd change should make a modest across the board improvement to test times.

File: build.yaml [hunk 1]
```diff
@@ -3645,12 +3694,12 @@ steps:
           --filter 'tags.items=batch2-agent AND labels.namespace={{ default_ns.name }}' \
           --format="table[no-heading](zone.basename(), name)" \
           --project {{ global.gcp_project }} \
-        | xargs -n2 -r sh -c 'gcloud -q compute instances delete --zone "$1" --project {{ global.gcp_project }} "$2" || true' argv0
+        | xargs -P4 -n2 -r sh -c 'gcloud -q compute instances delete --zone "$1" --project {{ global.gcp_project }} "$2" || true' argv0
```
Comment (daniel-goldstein) [inline build.yaml:3697]: This is probably fine, but have you tried doing, say, `P8`? I feel like we ought to be able to issue delete requests for all live instances at once and there could be many more than 4 across all the pools.

**Review by daniel-goldstein — CHANGES_REQUESTED** [review]
Awesome, just a couple nits.

**Comment by danking** [comment]
That huge spike in ready jobs has got to be a mistake. Where do we create a 1500 partition dataset?

Service backend [split 35](https://batch.hail.is/batches/7489026/jobs/197) is again the critical path. I tried pushing a commit which checkpoints the VCF. The main problem is probably just that our PC-Relate implementation is slow.
```
============================= slowest 50 durations =============================
240.09s call     hail/methods/relatedness/test_pc_relate.py::test_pc_relate_against_R_truth
168.03s call     hail/methods/test_pca.py::test_spectra_2[triplet0]
93.98s call     hail/vds/test_vds.py::test_truncate_reference_blocks
86.56s call     hail/backend/test_service_backend.py::test_tiny_driver_has_tiny_memory
83.82s call     hail/methods/test_qc.py::Tests::test_vep_grch38_against_dataproc
```

2a3fbd185e9255ed447dd80b983709d49c7a345e (45 minutes):

Screen Shot 2023-05-26 at 14 39 28


Screen Shot 2023-05-26 at 14 36 58

Screen Shot 2023-05-26 at 14 38 18
Screen Shot 2023-05-26 at 14 38 13

**Comment by danking** [comment]
Checkpointing the VCF made it, if anything, slower.
```
254.81s call     hail/methods/relatedness/test_pc_relate.py::test_pc_relate_against_R_truth
189.52s call     hail/methods/test_pca.py::test_spectra_2[triplet0]
95.37s call     hail/vds/test_vds.py::test_truncate_reference_blocks
92.80s call     hail/methods/test_qc.py::Tests::test_vep_grch38_against_dataproc
86.83s call     hail/backend/test_service_backend.py::test_tiny_driver_has_tiny_memory
68.86s call     hail/matrixtable/test_matrix_table.py::test_read_write_balding_nichols_model
```

**Comment by danking** [comment]
The test_hail_python tests have mysteriously disappeared.

**Comment by danking** [comment]
WIP, because there's too much chaos in CI right now. Let's squash the local backend issue first.

**Comment by danking** [comment]
I'm gonna pull these changes into a new PR

**Comment by danking** [comment]
https://github.com/hail-is/hail/pull/13128

**Pull request closed without merging**